### PR TITLE
Output flag

### PR DESCRIPTION
**Pull request by benjyz**
### What I did

add the flag "-o" to generate output files. this will write abi, bin and combined files. other options like ast, lll, opcodes need to be handled separately.

### How I did it

add a flag to argparser and write the files

### CODE REVIEW

File: vyper/cli/vyper_compile.py [hunk 1]
```diff
@@ -135,16 +138,33 @@ def _parse_args(argv):
...
+        for contract_data in compiled.values():
+            for data in contract_data.values():
+                if isinstance(data, (list, dict)):                    
+                    with open(args.output_path,"w") as f:
+                        f.write(json.dumps(data))
+                elif "bytecode" in contract_data.keys():
+                    with open(args.output_path,"w") as f:
+                        f.write(contract_data["bytecode"])
+                else:
+                    print("unrecognized output")
```
Comment (charles-cooper) [inline vyper/cli/vyper_compile.py:167]: Please throw an exception here instead

File: vyper/cli/vyper_compile.py [hunk 2]
```diff
@@ -135,16 +138,33 @@ def _parse_args(argv):
...
+            with open(args.output_path,"w") as f:
+                f.write(json.dumps(compiled))
+                return
+                
+        for contract_data in compiled.values():
+            for data in contract_data.values():
+                if isinstance(data, (list, dict)):                    
+                    with open(args.output_path,"w") as f:
+                        f.write(json.dumps(data))
+                elif "bytecode" in contract_data.keys():
```
Comment (charles-cooper) [inline vyper/cli/vyper_compile.py:163]: What is the purpose of this branch?

**Review by charles-cooper — CHANGES_REQUESTED** [review]
I think this PR will read a lot cleaner if you use a single branch, and just select a file handle depending on the option. If `-o` is set, set `f` to `args.output_path`. Otherwise, set it to `sys.stdout`.

File: vyper/cli/vyper_compile.py (hunk 2 above)
Comment (benjyz) [reply vyper/cli/vyper_compile.py:163]: this is for writing ABI files. would it be better to catch each case?

Comment (charles-cooper) [reply vyper/cli/vyper_compile.py:163]: Maybe I'm being dense here but what combination of flags activates this branch?

Comment (benjyz) [reply vyper/cli/vyper_compile.py:163]: thanks for the review, will make the suggest change so it's easier to read. I think in that case is pretty small just direct output to file or stdout and leave previous behavior.

**Comment by benjyz** [comment]
> I think this PR will read a lot cleaner if you use a single branch, and just select a file handle depending on the option. If `-o` is set, set `f` to `args.output_path`. Otherwise, set it to `sys.stdout`.

agreed, will do

**Comment by benjyz** [comment]
@charles-cooper changed the PR as you suggested. if two arguments for the flags are passed doesn't work well. in that case auto recognizing the format would be useful, i.e.

`vyper foo.vy -f bytecode,abi -o foo.txt` writes 1 file without breakline

**Review by fubuloubu — APPROVED** [review]

File: vyper/cli/vyper_compile.py [hunk 3]
```diff
@@ -135,16 +138,25 @@ def _parse_args(argv):
...
+
+    if args.output_path:
+        f = open(args.output_path,"w")
+    else:
+        f = sys.stdout
+
     if output_formats == ("combined_json",):
-        print(json.dumps(compiled))
+        f.write(json.dumps(compiled))
+        f.close()
```
Comment (charles-cooper) [inline vyper/cli/vyper_compile.py:149]: Please don't close the file handle. It could be stdout.

File: vyper/cli/vyper_compile.py [hunk 4]
```diff
@@ -135,16 +138,25 @@ def _parse_args(argv):
...
     for contract_data in compiled.values():
         for data in contract_data.values():
             if isinstance(data, (list, dict)):
-                print(json.dumps(data))
+                f.write(json.dumps(data))
             else:
-                print(data)
+                f.write(data)
+
+    f.close()
```
Comment (charles-cooper) [inline vyper/cli/vyper_compile.py:159]: No need to close the file handle here (and plus it could be stdout)

**Review by charles-cooper — CHANGES_REQUESTED** [review]

**Comment by charles-cooper** [comment]
In order to avoid managing the resource manually (i.e. inserting calls to `f.close()` before every escape path), i think a bit of refactoring is in order. To abstract which file handle is being used, you could use a helper function
```python
def cli_helper(f):
  ...
```
and then you would do something like
```python
if args.output_path:
  with open(...) as f:
    cli_helper(f)
else:
  cli_helper(f)
```

**Comment by charles-cooper** [comment]
Also, please use `print(..., file=f)` instead of `f.write()`. The latter is a break with existing semantics -- it breaks existing flags (like `-f ir`)

**Comment by benjyz** [comment]
@charles-cooper updated. I added end="" so that the output does not end with newline.

**Comment by charles-cooper** [comment]
> @charles-cooper updated. I added end="" so that the output does not end with newline.

I think we can accept this PR but without the `end=""`. That breaks the existing output format, which I am open to, but we should have an explicit discussion about that - either going through the VIP process or at least asking the major consumers of vyper if that's a breaking change for them first.

File: vyper/cli/vyper_compile.py [hunk 5]
```diff
@@ -54,6 +54,17 @@
 def _parse_cli_args():
     return _parse_args(sys.argv[1:])
 
+def cli_helper(f, output_formats, compiled):
```
Comment (charles-cooper) [inline vyper/cli/vyper_compile.py:57]:
```suggestion
def _cli_helper(f, output_formats, compiled):
```

File: vyper/cli/vyper_compile.py [hunk 6]
```diff
@@ -135,16 +149,13 @@ def _parse_args(argv):
...
 
-    for contract_data in compiled.values():
-        for data in contract_data.values():
-            if isinstance(data, (list, dict)):
-                print(json.dumps(data))
-            else:
-                print(data)
+    if args.output_path:
+        with open(args.output_path,"w") as f:
+            cli_helper(f, output_formats, compiled)
```
Comment (charles-cooper) [inline vyper/cli/vyper_compile.py:155]:
```suggestion
            _cli_helper(f, output_formats, compiled)
```

File: vyper/cli/vyper_compile.py [hunk 7]
```diff
@@ -135,16 +149,13 @@ def _parse_args(argv):
...
+    else:
+        f = sys.stdout
+        cli_helper(f, output_formats, compiled)
```
Comment (charles-cooper) [inline vyper/cli/vyper_compile.py:158]:
```suggestion
        _cli_helper(f, output_formats, compiled)
```

**Review by charles-cooper — APPROVED** [review]
Final nit: (style guide: functions private to a module should be prefixed with `_`)

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**